### PR TITLE
Add typescript on BO migrated pages and refacto grid system

### PR DESCRIPTION
**Pull request by NeOMakinG**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | We've no typing libs on the project, adding typescript would allow a lot of things : types, using TC39 new features like optional chaining...
| Type?             | improvement
| Category?         | BO
| BC breaks?        | yes
| Deprecations?     | no
| How to test?      | Build assets, it should build properly, and on this PR, grids should work as expected

# BC Break
It should be considered as a BC, as changing files extension is a BC and you can't import these files anymore without having a typescript-loader and tsconfig

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/23221)

### CODE REVIEW

File: admin-dev/themes/new-theme/js/types/grid/index.ts [hunk 1]
```diff
@@ -0,0 +1,3 @@
+export interface GridExtension {
+  extend: (grid: unknown) => void;
```
Comment (Progi1984) [inline admin-dev/themes/new-theme/js/types/grid/index.ts]: `unknown` ?

File: admin-dev/themes/new-theme/js/components/grid/extension/action/bulk/category/delete-categories-bulk-action-extension.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+export default class DeleteCategoriesBulkActionExtension {
+  /**
+   * Extend grid
+   *
+   * @param {Grid} grid
+   */
+  extend(grid: Grid): void {
+    grid
+      .getContainer()
+      .on('click', '.js-delete-categories-bulk-action', (event) => {
```
Comment (Progi1984) [inline admin-dev/themes/new-theme/js/components/grid/extension/action/bulk/category/delete-categories-bulk-action-extension.ts]: Could you move selector to a map file ?

File: admin-dev/themes/new-theme/js/components/grid/extension/action/bulk/category/delete-categories-bulk-action-extension.ts [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        const submitUrl = $(event.currentTarget).data('categories-delete-url');
+
+        const $deleteCategoriesModal = $(
+          `#${grid.getId()}_grid_delete_categories_modal`,
+        );
+        $deleteCategoriesModal.modal('show');
+
+        $deleteCategoriesModal.on(
+          'click',
+          '.js-submit-delete-categories',
```
Comment (Progi1984) [inline admin-dev/themes/new-theme/js/components/grid/extension/action/bulk/category/delete-categories-bulk-action-extension.ts]: Could you move selector to a map file ?

File: admin-dev/themes/new-theme/js/components/grid/extension/action/bulk/category/delete-categories-bulk-action-extension.ts [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+          () => {
+            const $checkboxes = grid
+              .getContainer()
+              .find('.js-bulk-action-checkbox:checked');
```
Comment (Progi1984) [inline admin-dev/themes/new-theme/js/components/grid/extension/action/bulk/category/delete-categories-bulk-action-extension.ts]: Could you move selector to a map file ?

File: admin-dev/themes/new-theme/js/components/grid/extension/action/bulk/customer/delete-customers-bulk-action-extension.ts [hunk 5]
```diff
@@ -22,56 +22,55 @@
...
    *
    * @param {Grid} grid
    */
-  extend(grid) {
-    grid.getContainer().on('click', '.js-delete-customers-bulk-action', (event) => {
-      event.preventDefault();
+  extend(grid: Grid): void {
+    grid
+      .getContainer()
+      .on('click', '.js-delete-customers-bulk-action', (event) => {
```
Comment (Progi1984) [inline admin-dev/themes/new-theme/js/components/grid/extension/action/bulk/customer/delete-customers-bulk-action-extension.ts]: Could you move selector to a map file ?

File: admin-dev/themes/new-theme/js/components/router.ts [hunk 6]
```diff
@@ -39,7 +41,8 @@ const {$} = window;
  *    options:
  *      expose: true
  * `
- * And run `bin/console fos:js-routing:dump --format=json --target=admin-dev/themes/new-theme/js/fos_js_routes.json`
+ * And run `bin/console fos:js-routing:dump --format=json
+ * --target=admin-dev/themes/new-theme/js/fos_js_routes.json`
```
Comment (Progi1984) [inline admin-dev/themes/new-theme/js/components/router.ts]:
```suggestion
 * And run `bin/console fos:js-routing:dump --format=json --target=admin-dev/themes/new-theme/js/fos_js_routes.json`
```

File: admin-dev/themes/new-theme/js/components/grid/extension/submit-grid-action-extension.ts [hunk 7]
```diff
@@ -22,23 +22,20 @@
...
-  }
-
-  extend(grid) {
-    grid.getHeaderContainer().on('click', '.js-grid-action-submit-btn', (event) => {
-      this.handleSubmit(event, grid);
-    });
+  extend(grid: Grid): void {
+    grid
+      .getHeaderContainer()
+      .on('click', '.js-grid-action-submit-btn', (event: JQueryEventObject) => {
```
Comment (Progi1984) [inline admin-dev/themes/new-theme/js/components/grid/extension/submit-grid-action-extension.ts]: Could you selector to a map file?

File: admin-dev/themes/new-theme/js/components/grid/extension/submit-bulk-action-extension.ts [hunk 8]
```diff
@@ -31,21 +32,17 @@ const {$} = window;
...
    * @param {Grid} grid
    */
-  extend(grid) {
-    grid.getContainer().on('click', '.js-bulk-action-submit-btn', (event) => {
-      this.submit(event, grid);
-    });
+  extend(grid: Grid): void {
+    grid
+      .getContainer()
+      .on('click', '.js-bulk-action-submit-btn', (event: JQueryEventObject) => {
```
Comment (Progi1984) [inline admin-dev/themes/new-theme/js/components/grid/extension/submit-bulk-action-extension.ts]: Could you move selector to a map file?

File: admin-dev/themes/new-theme/js/components/grid/extension/reload-list-extension.ts [hunk 9]
```diff
@@ -32,9 +34,11 @@ export default class ReloadListExtension {
...
    * @param {Grid} grid
    */
-  extend(grid) {
-    grid.getHeaderContainer().on('click', '.js-common_refresh_list-grid-action', () => {
-      window.location.reload();
-    });
+  extend(grid: Grid): void {
+    grid
+      .getHeaderContainer()
+      .on('click', '.js-common_refresh_list-grid-action', () => {
```
Comment (Progi1984) [inline admin-dev/themes/new-theme/js/components/grid/extension/reload-list-extension.ts]: Could you move selector to a map file?

File: admin-dev/themes/new-theme/js/components/grid/extension/position-extension.js [hunk 10]
```diff
@@ -45,19 +45,29 @@ export default class PositionExtension {
...
-      },
-      function () {
-        $(this).closest('tr').removeClass('hover');
-      },
-    );
+    grid
+      .getContainer()
+      .find('.js-grid-table')
+      .tableDnD({
+        onDragClass: 'position-row-while-drag',
```
Comment (Progi1984) [inline admin-dev/themes/new-theme/js/components/grid/extension/position-extension.js]: Could you move selectors to a map file?

File: admin-dev/themes/new-theme/js/components/grid/extension/filters-submit-button-enabler-extension.ts [hunk 11]
```diff
@@ -32,13 +34,15 @@ export default class FiltersSubmitButtonEnablerExtension {
...
 
-    $filtersRow.find('input:not(.js-bulk-action-select-all), select').on('input dp.change', () => {
-      $filtersRow.find('.grid-search-button').prop('disabled', false);
-      $filtersRow.find('.js-grid-reset-button').prop('hidden', false);
-    });
+    $filtersRow
+      .find('input:not(.js-bulk-action-select-all), select')
+      .on('input dp.change', () => {
+        $filtersRow.find('.grid-search-button').prop('disabled', false);
+        $filtersRow.find('.js-grid-reset-button').prop('hidden', false);
```
Comment (Progi1984) [inline admin-dev/themes/new-theme/js/components/grid/extension/filters-submit-button-enabler-extension.ts]: Could you move selectors to a map file?

File: admin-dev/themes/new-theme/js/components/grid/extension/export-to-sql-manager-extension.ts [hunk 12]
```diff
@@ -34,13 +37,15 @@ export default class ExportToSqlManagerExtension {
...
-    );
+  extend(grid: Grid): void {
+    grid
+      .getHeaderContainer()
+      .on('click', GridMap.actions.showQuery, () => this.onShowSqlQueryClick(grid),
+      );
+    grid
+      .getHeaderContainer()
+      .on('click', GridMap.actions.exportQuery, () => this.onExportSqlManagerClick(grid),
+      );
```
Comment (Progi1984) [inline admin-dev/themes/new-theme/js/components/grid/extension/export-to-sql-manager-extension.ts]:
```suggestion
      .on('click', GridMap.actions.exportQuery, () => this.onExportSqlManagerClick(grid));
```

File: admin-dev/themes/new-theme/js/components/grid/extension/export-to-sql-manager-extension.ts [hunk 13]
```diff
@@ -34,13 +37,15 @@ export default class ExportToSqlManagerExtension {
...
-    grid.getHeaderContainer().on(
-      'click',
-      '.js-common_export_sql_manager-grid-action',
-      () => this.onExportSqlManagerClick(grid),
-    );
+  extend(grid: Grid): void {
+    grid
+      .getHeaderContainer()
+      .on('click', GridMap.actions.showQuery, () => this.onShowSqlQueryClick(grid),
+      );
```
Comment (Progi1984) [inline admin-dev/themes/new-theme/js/components/grid/extension/export-to-sql-manager-extension.ts]:
```suggestion
      .on('click', GridMap.actions.showQuery, () => this.onShowSqlQueryClick(grid));
```

**Review by Progi1984 — COMMENTED** [review]
Some feedbacks

File: admin-dev/themes/new-theme/js/components/grid/extension/position-extension.ts [hunk 14]
```diff
@@ -23,41 +23,64 @@
...
+      .getContainer()
+      .find(GridMap.gridTable)
+      .tableDnD({
+        onDragClass: GridMap.onDragClass,
+        dragHandle: GridMap.dragHandler,
+        onDrop: (table: HTMLElement, row: HTMLElement) => this.handlePositionChange(row),
+      });
+    grid
+      .getContainer()
+      .find('.js-drag-handle')
```
Comment (Progi1984) [inline admin-dev/themes/new-theme/js/components/grid/extension/position-extension.ts:71]:
```suggestion
      .find(GridMap.dragHandler)
```

**Review by Progi1984 — DISMISSED** [review]

**Review by PierreRambaud — DISMISSED** [review]

**Comment by NeOMakinG** [comment]
@PierreRambaud @Progi1984 I rebased due to the 1.7.8 merge into develop

File: admin-dev/themes/new-theme/js/types/global.d.ts [hunk 15]
```diff
@@ -0,0 +1,36 @@
+interface Window {
```
Comment (PierreRambaud) [inline admin-dev/themes/new-theme/js/types/global.d.ts:1]: Why ` .d`  in the name of the file, can you add a little comment at the top of this file to explain what's the purpose?

**Review by PierreRambaud — APPROVED** [review]

**Review by Progi1984 — APPROVED** [review]

**Comment by PierreRambaud** [comment]
PIng @Robin-Fischer-PS What about this one, could we just execute the full UI tests suite, I'm pretty sure we will find some problem in the future, it's like updating PHP :sweat_smile:

**Comment by hibatallahAouadni** [comment]
Ping @PierreRambaud with this PR I encounter a lot of regressions:

**Order page (desktop) :**

image

**Invoices page:**

image

**Product page:**

image

etc...

**Comment by hibatallahAouadni** [comment]
Hello @NeOMakinG 

Tested again after the rebase, everything looks good but these two issues:

**- Order page (desktop):**

image

**- Database > DB Backup (mobile)**

image

Please check and feedback.
Thanks!

**Comment by NeOMakinG** [comment]
Hey @hibatallahAouadni thanks a lot for testing, it looks like these issues are real, but not related to this PR

This PR is all about grids

**Comment by hibatallahAouadni** [comment]
Great then LGTM :heavy_check_mark:

**Comment by PierreRambaud** [comment]
Thank you @NeOMakinG

**Pull request merged**